### ISSUE TITLE
DataProcessor - implementation

### ISSUE THREAD

**Issue by szymonkups** [ckeditor/ckeditor5-core#171]
I had a face to face talk with @Reinmar about data processor implementation. 
First we need to create a `DataProcessor` "interface", with two basic methods:
* **toDom** - should take an input data and convert it to DOM document,
* **toData** - should take a DOM document as parameter and convert it back to data format.

Next step is to create `HtmlDataProcessor`. It will the most basic, and probably most commonly used Data Processor.
* **toDom** method should take a HTML string as parameter and convert it to DOM Document, uses [DOMParser](https://developer.mozilla.org/en-US/docs/Web/API/DOMParser)
* **toData** method should take DOM document as input and return HTML string, uses `HtmlWriter`.

**HtmlWriter** should be introduced. It should be able to convert DOM Document to string format. We decided that it should be separated from `HtmlDataProcessor` as it might be useful for other data processors' implementations. It should provide options to control output HTML string, like adding indentations, converting entities, sorting attributes etc. 

#### Notes
1. Some data processors might require a configuration, it could be stored under `editor.config.dataProcessor` and passed to data processor's constructor.
2. Asynchronous data processors will be introduced later.

**Comment by Reinmar** [comment]
As for HtmlWriter, I don't think that we necessarily must work on it right now, as this will be a bigger piece of code. We can for now use some `BasicHtmlWriter` printing out the `innerHTML` string.

**Comment by fredck** [comment]
+1

**Comment by szymonkups** [comment]
I stared using `DOMParser` for HTML to DOM conversion. `DOMParser.parseFromString` method returns whole `HTMLDocument` (including `html`, `head` and `body` tags):

```js
    const parser = new DOMParser();
    const htmlString = '<p>Link to <a href="http://ckeditor.com">CKEditor</a></p>';

    // returns HTMLDocument
    const document = parser.parseFromString( htmlString, 'text/html' ); 
```

We decided that `toDom()` method will return `DocumentFragment` that will include all nodes from document's body returned by `DOMParser`.

**Comment by Reinmar** [comment]
For now we can return the contents of that document `<body>`, assuming that by default this data processor does not support parsing whole documents. We can later add this feature as an option. if we'll ever decide to implement CKE4's full page editing... or when the clipboard implementation will need it, because sometimes we may need to access the meta tags of the content available there.

**Comment by Reinmar** [comment]
Implemented in #178.
